Naming

m7pr opened the pull request:
Changed names

- `teal_reporter@report` ->  `teal_reporter@teal_card`
- `teal.reporter::doc()` ->  `teal.reporter::teal_card()`
- `teal.reporter::report()` ->  `teal.reporter::teal_card()`

So now you can use

```r
data <- teal.reporter::teal_report()  # teal_report() class
data@teal_card                        # teal_card class
teal.reporter::teal_card(data)        # teal_card class
```

Also `teal_card` is the creator, the getter and the setter

```r
data <- teal.reporter::teal_report()
teal_card(data) <- teal_card()       # setter + creator
teal_card(data)                      # getter
data@teal_card                       # direct slot access
```

Companion to:
- https://github.com/insightsengineering/teal/pull/1542
- https://github.com/insightsengineering/teal.modules.general/pull/885
- https://github.com/insightsengineering/teal.code/pull/256

m7pr commented on R/teal_report-report.R:
```diff
@@ -1,22 +1,19 @@
-#' @return `teal_report`
+#' @param value (`teal_document`)
+#'
+#' @return The document of the `teal_report` object.
```
?
```suggestion
#' @return The document slot of the `teal_report` object.
```

gogonzo reviewed (commented):
Can we have the same function name for get/set and create? Like with `join_keys`?  

```r
td <- teal.reporter::teal_data()
join_keys(td) <- join_keys()
join_keys(td)


obj <- teal.reporter::teal_report()
document(obj) <- teal_document()
document(obj)
```

-----

There are some `doc(` remaining in `teal` and `teal.reporter`

m7pr commented on R/teal_report-class.R:
```diff
@@ -103,10 +103,35 @@ as.teal_report <- function(x) {
+#' @export
+`document<-` <- function(x, value) {
+  x@document <- as_teal_document(value)
+  x
```
Those are not the same

```
`document<-`()
```

just sets the `@document` field, 
where 
```
append_document()
```
extends the `@document` field

gogonzo replied:
I'd add `append_document` in another PR. PR is about renaming, `append_document` is another improvement which can start a new discussion

m7pr commented:
Based on this comment https://github.com/insightsengineering/teal.reporter/pull/334#issuecomment-2939212626 and this https://github.com/insightsengineering/teal.reporter/pull/334#pullrequestreview-2895944416 (do not use `document()` function, and have the same name for getter and setter, I think we can go with 

`teal_document()`


```r
data<- teal.reporter::teal_report()
teal_document(data) <- teal_document()
teal_document(data)
data@document
```

m7pr commented on R/teal_report-class.R:
```diff
@@ -103,10 +103,35 @@ as.teal_report <- function(x) {
+#' @param x Object to convert to teal_document
+#' @return A teal_document object
+#' @export
+as_teal_document <- function(x) {
```
changed in here https://github.com/insightsengineering/teal.reporter/pull/334/commits/2ad9d24afa944de1fc6197cd9f605a6c02b010ba

m7pr commented:
I also think I need to update tests. Also there is a remaining called `edit_doc` in tests that does not have a definition

m7pr commented on R/doc.R:
```diff
@@ -135,25 +135,25 @@ metadata.ReportCard <- function(object, which = NULL) {
 #' @export
-edit_doc <- function(x, modify = NULL, append = NULL, after = length(x)) {
-  checkmate::assert_class(x, "doc")
+edit_teal_document <- function(x, modify = NULL, append = NULL, after = length(x)) {
```
@gogonzo @averissimo do we really need `edit_teal_document`? It's quite complex operator that aims to do `x[i]` (substract) and append at the same time. Maybe having `x[i]` and `append_document()` is enough?

averissimo replied:
I agree, it's a complex method that can be achieved in other ways.

The only limitation is that we can't be directly append an element to start without ensuring it's a `teal_document`

``` r
pkgload::load_all("../teal.reporter")

doc1 <- teal_document("## a title", "a paragraph", ggplot2::qplot(1:10, 1:10))

c("pre-title", doc1) |> 
  checkmate::expect_class("teal_document")
#> Error: Check on 'c("pre-title", doc1)' failed: Must inherit from class 'teal_document', but has class 'list'

c(teal_document("pre-title"), doc1) |> 
  checkmate::expect_class("teal_document")

edit_teal_document(doc1, append = "pre-title", after = 0) |> 
  checkmate::expect_class("teal_document")
```

<sup>Created on 2025-06-05 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

m7pr commented:
@gogonzo @averissimo I updated the PR.
Also I updated the initial message. Now we don't have `document()` function, but we have `teal_document()` that is a creator, getter and setter. I also updated tests and documentation,

m7pr commented on R/teal_document.R:
```diff
@@ -1,74 +1,114 @@
-doc <- function(...) {
-  objects <- list(...)
-  structure(objects, class = c("doc"))
+teal_document <- function(x, ...) {
```
ouch, I had it before the merge with feature branch. somehow deleted handling the missing x
fixed in here https://github.com/insightsengineering/teal.reporter/pull/334/commits/eb5e9b3b1a1641392906163f73c43380daf0e22a

m7pr commented:
> WDYT @averissimo @m7pr isn't teal.reporter::teal_document(data) too long? I also think that the slotname could have the same name as function used to get/set.

@gogonzo @averissimo about the name for slot and the setter/getter. Below some ideas instead of

`@document + teal_document()`


`@items/item + teal.report::item()`
`@pages/page + teal.report::page()`
`@content + teal.report::entry()`

averissimo commented:
> > WDYT @averissimo @m7pr isn't teal.reporter::teal_document(data) too long? I also think that the slotname could have the same name as function used to get/set.
> 
> @gogonzo @averissimo about the name for slot and the setter/getter. Below some ideas instead of
> 
> `@document + teal_document()`
> 
> `@items/item + teal.report::item()` `@pages/page + teal.report::page()` `@content + teal.report::entry()`

How about we use `@card` + `teal.reporter::card()`

The downside is that it's a concept that already exists with `ReporterCard`, but it's short and kinda to the point.

Otherwise, my vote goes to `@page` the others are too broad

m7pr commented:
Whoah, I love `@card` and `teal.reporter::card()` :)

m7pr commented:
Hey, made progress with `@card and teal.reporter::card()` and updated the initial message.
In the end it's

```r
data <- teal.reporter::teal_report()
card(data) <- card()  # setter + creator
card(data)            # getter
data@card             # direct slot access
```

m7pr commented:
Also changed `cache` parameter to `keep_output` and attribute name from `cache` to `output`

```r
> library(teal.reporter)
> data <- teal_report() 
> library(teal.code)
> eval_code(data, "x <- 5", keep_output = TRUE) -> data2
> data2@code  
$`1799856228`
[1] "x <- 5"
attr(,"output")
[1] 5
attr(,"dependency")
[1] "x"  "<-"
```

gogonzo commented on R/teal_report-class.R:
```diff
@@ -80,13 +81,14 @@ setMethod(
+                         code = character(0),
+                         join_keys = teal.data::join_keys()) {
+  if (is.null(card)) card <- teal.reporter::card()
   methods::new(
     "teal_report",
     .xData = list2env(list(...)),
-    report = report,
+    card = card,
```
👎 
```r
teal_report()
# Error in .local(.Object, ...) : 
#  Assertion on '.xData' failed: Must be an environment, not 'list'.
```

```suggestion
  methods::new(
    "teal_report",
    .xData = list2env(list(...)),
    card = if (is.null(card)) card(),
```

m7pr replied:
That's what I get on my end

```r
Running "C:/PROGRA~1/R/R-devel/bin/x64/Rcmd.exe" INSTALL \
  "C:\Users\kosinsm4\AppData\Local\Temp\RtmpwFhHWK/teal.reporter_0.4.0.9004.tar.gz" \
  --install-tests
* installing to library 'C:/Users/kosinsm4/AppData/Local/R/win-library/4.4'
* installing *source* package 'teal.reporter' ...
** using staged installation
** R
** inst
** tests
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
*** copying figures
** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (teal.reporter)
> library(teal.reporter)
> teal_report()  
✅︎ code verified
<environment: 0x000002a06de75470> 🔒
Parent: <environment: package:teal.reporter> 
```

gogonzo reviewed (changes requested):
Example app just died, `teal_report()` throws an error. Looks like commit was lacking sanity checks.

m7pr commented:
Hi @gogonzo , what do you mean by `Example app just died`

I just run 

# all 4 packages on branch `naming@teal_reportable`

```r
devtools::load_all("teal.code")
devtools::load_all("teal.data")
devtools::load_all("teal.reporter")
devtools::load_all("teal")

app <- init(
  data = within(
    teal_data(),
    {
      new_iris <- transform(iris, id = seq_len(nrow(iris)))
      new_mtcars <- transform(mtcars, id = seq_len(nrow(mtcars)))
    }
  ),
  modules = modules(
    module(
      label = "data source",
      server = function(input, output, session, data) {},
      ui = function(id, ...) tags$div(p("information about data source")),
      datanames = "all"
    ),
    example_module(label = "example teal module"),
    module(
      "Iris Sepal.Length histogram",
      server = function(input, output, session, data) {
        output$hist <- renderPlot(
          hist(data()[["new_iris"]]$Sepal.Length)
        )
      },
      ui = function(id, ...) {
        ns <- NS(id)
        plotOutput(ns("hist"))
      },
      datanames = "new_iris"
    )
  ),
  filter = teal_slices(
    teal_slice(dataname = "new_iris", varname = "Species"),
    teal_slice(dataname = "new_iris", varname = "Sepal.Length"),
    teal_slice(dataname = "new_mtcars", varname = "cyl"),
    exclude_varnames = list(new_iris = c("Sepal.Width", "Petal.Width")),
    module_specific = TRUE,
    mapping = list(
      `example teal module` = "new_iris Species",
      `Iris Sepal.Length histogram` = "new_iris Species",
      global_filters = "new_mtcars cyl"
    )
  )
)
if (interactive()) {
  shinyApp(app$ui, app$server)
}

```

And the app starts and the reporter works (adds cards to the report, views them in the previewer)


https://github.com/user-attachments/assets/3887c6b7-ef74-4661-b53d-d7ffb6dcfc2d

gogonzo commented on tests/testthat/test-LoadReporterModule.R:
```diff
@@ -106,3 +106,36 @@ testthat::test_that("report_load_srv - fail to load a reporter because of differ
   checkmate::expect_multi_class(report_load_ui("sth"), c("shiny.tag.list", "shiny.tag"))
 })
+
+testthat::test_that("LoadReporterModule works", {
```
This test looks like it was improvised. There is no such module. Reporter is not saved directly through saveRDS but there is a method in the Reporter object.
Please also be more specific in test name. Should rather be `report_load/ui/srv` restores reporter from saved bundle.

m7pr commented on tests/testthat/helper-Reporter.R:
```diff
@@ -64,3 +64,33 @@ test_reporter <- function(card1 = test_card1(), card2 = test_card2(), ...) {
   reporter
 }
+
+create_test_reporter <- function() {
```
@averissimo we now have 2 `create_test_reporter`. Would you be able to help determine which one we should keep?
![image](https://github.com/user-attachments/assets/42ffad0b-7690-454b-97a8-8318fbfd9a59)

m7pr commented on tests/testthat/helper-Reporter.R:
```diff
@@ -64,3 +64,33 @@ test_reporter <- function(card1 = test_card1(), card2 = test_card2(), ...) {
   reporter
 }
+
+create_test_reporter <- function() {
+  testthat::skip_if_not_installed("ggplot2")
+  library(ggplot2)
+
+  reporter <- Reporter$new()
+
+  # Create a plot
+  plot <- ggplot(iris) +
+    geom_point(aes(x = Sepal.Length, y = Sepal.Width))
+
+  # Create a table
+  table_res <- summary(iris)
+
+  # Create a table with a caption
+  table_res2 <- summary(iris)
+  attr(table_res2, "caption") <- "Summary of iris dataset"
+
+  # Create a new card with a plot
+  new_card <- card("## Header 2 text", "A paragraph of default text", plot)
+  metadata(new_card, "title") <- "Card with plot"
+  reporter$append_cards(new_card)
+
+  # Create a new card with a table
+  new_card <- card("## Header 2 text", "A paragraph of default text", table_res2, iris)
+  metadata(new_card, "title") <- "Card with table"
+  reporter$append_cards(new_card)
+
+  reporter
+}
```
```suggestion
```

averissimo commented on R/teal_report-within.R:
```diff
@@ -0,0 +1,9 @@
+setMethod(
+  "within",
+  signature = c(data = "teal_report"),
+  function(data, expr, keep_output = FALSE, ...) {
+    expr <- substitute(expr)
+    code <- deparse(expr)
+    eval_code(data, code = code, keep_output = keep_output, ...)
+  }
+) 
```
Why do we need this method?

averissimo replied:
The only method that we need to overwrite is `eval_code` as that will be dispatched once `teal.code::within` calls it.

If we really need this we should make it very similar to `teal.code`'s within definition

``` r
pkgload::load_all("/home/averissimo/work/roche🟦/packages📦/teal.code")
#> ℹ Loading teal.code
pkgload::load_all("/home/averissimo/work/roche🟦/packages📦/teal.data")
#> ℹ Loading teal.data
pkgload::load_all("/home/averissimo/work/roche🟦/packages📦/teal.reporter")
#> ℹ Loading teal.reporter

td <- teal_data()
td <- td |> within(iris <- iris)

tr <- teal.reporter::as.teal_report(td)

tr |> within({
  1+1
})
#> <qenv.error: object 'expr' not found 
#>  when evaluating qenv code:
#> expr>
```

<sup>Created on 2025-06-06 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

m7pr commented:
After the talk with @averissimo we decided to change `card` field and function name into `teal_card` because `card` is used by `cards` package. I think prepending objects and function names with `teal_` is the only possibility to over-come namespacing collisions and clearly states its the function from this family.

So next to `teal_data`, `teal_slices` and `teal_transform` we now have `teal_card`.

averissimo commented on R/teal_report-class.R:
```diff
@@ -26,16 +26,15 @@ setOldClass("doc")
-#'  Used internally. See [`teal.data::verify()`] for more details.
-#' @slot report (`doc`)
-#'
+#'  Used internally. See [`verify()`] for more details.
```
```suggestion
#'  Used internally. See [`data::verify()`] for more details.
```

averissimo replied:
We don't re-export `verify()`

averissimo commented on R/teal_report-class.R:
```diff
@@ -78,13 +78,14 @@ setMethod(
-                        report = doc(),
-                        code = character(0),
-                        join_keys = teal.data::join_keys()) {
+                         teal_card = NULL,
+                         code = character(0),
+                         join_keys = teal.data::join_keys()) {
```
1 too many spaces?

```suggestion
                        teal_card = NULL,
                        code = character(0),
                        join_keys = teal.data::join_keys()) {
```

averissimo commented on R/toHTML.R:
```diff
@@ -31,9 +31,9 @@ toHTML.default <- function(x, ...) {
 }
 
-#' @method .toHTML doc
+#' @method .toHTML card
```
```suggestion
#' @method .toHTML teal_card
```

averissimo commented on R/toHTML.R:
```diff
@@ -31,9 +31,9 @@ toHTML.default <- function(x, ...) {
+#' @method .toHTML card
 #' @keywords internal
-.toHTML.doc <- function(x, ...) {
+.toHTML.card <- function(x, ...) {
```
```suggestion
.toHTML.teal_card <- function(x, ...) {
```
